Refactored to use upcoming EventManager v3

weierophinney opened the pull request:
Updates all code to work with the upcoming v3 of EventManager. In particular:
- Updated the EventManager factory to inject the SharedEventManager at instantiation, if that service is present.
- Updated the EventManager initializer to inject not only if no EM instance is present, but also if the current EM instance has no shared manager present.  The latter is a good indication that the EM instance was lazy-loaded by the getter.
- Updated the Application constructor to pass arguments. This was necessary due to a strange circular dependency with the changes to the initializer. It's also better DI.
- Fixed all `attachAggregate()` calls to use `$aggregate->attach()`, and ensured there were no calls to `$events->attach()` that were attaching aggregates.
- Reviewed all `trigger()` calls for appropriate signatures, updating those that needed it. In many cases, an additional call to `$event->setName()` was also required.

This can likely be merged to develop sooner rather than later, as it represents a work-in-progress towards the updated MVC.

weierophinney committed:
Refactored to use upcoming EventManager v3

Updates all code to work with the upcoming v3 of EventManager. In particular:

- Updated the EventManager factory to inject the SharedEventManager at
  instantiation, if that service is present.
- Updated the EventManager initializer to inject not only if no EM instance is
  present, but also if the current EM instance has no shared manager present.
  The latter is a good indication that the EM instance was lazy-loaded by the
  getter.
- Updated the Application constructor to pass arguments. This was necessary due
  to a strange circular dependency with the changes to the initializer. It's
  also better DI.
- Fixed all `attachAggregate()` calls to use `$aggregate->attach()`, and ensured
  there were no calls to `$events->attach()` that were attaching aggregates.
- Reviewed all `trigger()` calls for appropriate signatures, updating those that
  needed it. In many cases, an additional call to `$event->setName()` was also
  required.

This can likely be merged to develop sooner rather than later, as it represents
a work-in-progress towards the updated MVC.

weierophinney commented:
Pinging @zendframework/community-review-team and @ezimuel — would love some review. :smile:

danizord commented on src/Application.php:
```diff
@@ -103,15 +104,18 @@ class Application implements
+    public function __construct(
+        $configuration,
+        ServiceManager $serviceManager,
+        EventManagerInterface $events,
```
Should we provide defaults to avoid BC break?

Ocramius replied:
Absolutely agree here.

gianarb replied:
This is a BC break but we are working to mvc 3.0 if it is a GOOD BC break this is a good time to merge it

Ocramius replied:
@gianarb it is kind of a big one, tbh, and it can be avoided

weierophinney replied:
Putting a note here as threads got broken up by me answering via email.

This change was for a reason, which I stated in the PR summary: I ended up with a circular reference issue with the ApplicationFactory + EventManagerAware initializer that resulted in invalid EventManager state. The result was that the EM instance injected no longer had the default listeners injected, which broke the entire thing.

As I noted in another comment elsewhere: hardly anybody directly instantiates the `Application` instance. The skeleton application pulls it from the service manager. As such, we can call this change out in the migration documentation for the very small percentage of users who are manually instantiating.

Ocramius commented on src/Controller/AbstractController.php:
```diff
@@ -109,13 +109,14 @@ public function dispatch(Request $request, Response $response = null)
         $this->response = $response;
 
         $e = $this->getEvent();
```
While at it, maybe rename it `$event`? ;-)

gianarb replied:
:+1:

weierophinney replied:
Out of scope for this PR; I tried very hard only to update the code necessary to get the EM refactor working with it. If you feel strongly, open new PRs.

Ocramius commented on src/Controller/ControllerManager.php:
```diff
@@ -73,10 +74,10 @@ public function injectControllerDependencies($controller, ServiceLocatorInterfac
             // will always pass.
             $events = $controller->getEventManager();
-            if (!$events instanceof EventManagerInterface) {
+            if (! $events instanceof EventManagerInterface
```
`if (! $events) {` is sufficient here, as per `EventManagerAwareInterface`

Ocramius replied:
Urgh, wait, the interface defines no getters... https://github.com/zendframework/zend-eventmanager/blob/a6c747fe59658fb62feb23613d7ea22146b6b611/src/EventManagerAwareInterface.php#L15-L24

weierophinney replied:
Yes it does! It inherits from EventsCapableInterface, which defines the
getter.
On Oct 1, 2015 6:37 AM, "Marco Pivetta" notifications@github.com wrote:

> In src/Controller/ControllerManager.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40903723:
> 
> > @@ -73,10 +74,10 @@ public function injectControllerDependencies($controller, ServiceLocatorInterfac
> >              // is why the shared EM injection needs to happen; the conditional
> >              // will always pass.
> >              $events = $controller->getEventManager();
> > -            if (!$events instanceof EventManagerInterface) {
> > -            if (! $events instanceof EventManagerInterface
> 
> Urgh, wait, the interface defines no getters...
> https://github.com/zendframework/zend-eventmanager/blob/a6c747fe59658fb62feb23613d7ea22146b6b611/src/EventManagerAwareInterface.php#L15-L24
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40903723.

Ocramius replied:
:+1:

Ocramius commented on src/Controller/Plugin/Forward.php:
```diff
@@ -193,7 +193,7 @@ protected function detachProblemListeners(SharedEvents $sharedEvents)
                     foreach ($classArray as $class) {
                         if ($currentCallback instanceof $class) {
-                            $sharedEvents->detach($id, $currentEvent);
+                            $sharedEvents->detach($currentEvent, $id);
```
This is a super-subtle bc break. Is it possible to revert this in the EVM instead?

weierophinney replied:
No. In the new version, the second argument is optional, the first required.
On Oct 1, 2015 6:25 AM, "Marco Pivetta" notifications@github.com wrote:

> In src/Controller/Plugin/Forward.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40902934:
> 
> > @@ -193,7 +193,7 @@ protected function detachProblemListeners(SharedEvents $sharedEvents)
> > 
> > ```
> >                  foreach ($classArray as $class) {
> >                      if ($currentCallback instanceof $class) {
> > ```
> > -                            $sharedEvents->detach($id, $currentEvent);
> > -                            $sharedEvents->detach($currentEvent, $id);
> 
> This is a super-subtle bc break. Is it possible to revert this in the EVM
> instead?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40902934.

Ocramius replied:
I see. Still very very very annoying BC break

weierophinney replied:
It's a rarely used feature, TBH; there are very few use cases currently for detaching listeners, much less shared listeners, and the `Forward` plugin here is literally the only place I've seen it in the ZF2 or Apigility code bases.

Though with React, I suspect it will becomes more common. Those users will adopt v3 anyways, as it's faster. :smile:

Ocramius commented on src/Application.php:
```diff
@@ -350,8 +359,12 @@ protected function completeRequest(MvcEvent $event)
+
+        $event->setName(MvcEvent::EVENT_FINISH);
+        $events->triggerEvent($event);
         return $this;
```
Newline before this line (keeps consistent with the changes above)

weierophinney replied:
Noted, and staged locally to push with other changes.

Ocramius commented on src/Application.php:
```diff
@@ -311,14 +318,16 @@ public function run()
-            $events->trigger(MvcEvent::EVENT_FINISH, $event);
+            $events->triggerEvent($event);
             $this->response = $response;
             return $this;
```
Newline before this line (keeps consistent with the changes above)

Ocramius commented on src/DispatchListener.php:
```diff
@@ -92,12 +93,14 @@ public function onDispatch(MvcEvent $e)
+            $e->setName(MvcEvent::EVENT_DISPATCH_ERROR);
+            $e->setError($application::ERROR_EXCEPTION);
+            $e->setController($controllerName);
+            $e->setControllerClass(get_class($controller));
```
Can this be part of the event class instead?

weierophinney replied:
That's out of scope for this particular refactor.
On Oct 1, 2015 6:34 AM, "Marco Pivetta" notifications@github.com wrote:

> In src/DispatchListener.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40903363:
> 
> > @@ -92,12 +93,14 @@ public function onDispatch(MvcEvent $e)
> >          try {
> >              $return = $controller->dispatch($request, $response);
> >          } catch (\Exception $ex) {
> > -            $e->setError($application::ERROR_EXCEPTION)
> > -                  ->setController($controllerName)
> > -                  ->setControllerClass(get_class($controller))
> > -                  ->setParam('exception', $ex);
> > -            $results = $events->trigger(MvcEvent::EVENT_DISPATCH_ERROR, $e);
> > -            $return = $results->last();
> > -            $e->setName(MvcEvent::EVENT_DISPATCH_ERROR);
> > -            $e->setError($application::ERROR_EXCEPTION);
> > -            $e->setController($controllerName);
> > -            $e->setControllerClass(get_class($controller));
> 
> Can this be part of the event class instead?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40903363.

Ocramius replied:
Alright, will raise it as an issue after we get to the end with this PR

Ocramius commented on src/DispatchListener.php:
```diff
@@ -92,12 +93,14 @@ public function onDispatch(MvcEvent $e)
+            $e->setParam('exception', $ex);
+
+            $results = $events->triggerEvent($e);
+            $return  = $results->last();
```
Can chain here

weierophinney replied:
Chain which? I purposely didn't chain the setters for the event, as the `EventInterface` no longer defines fluent setters (though `MvcEvent` still does). Or did you mean this?

``` php
$return = $events->triggerEvent($e)->last();
```

Ocramius commented on src/DispatchListener.php:
```diff
@@ -211,12 +215,13 @@ protected function marshalBadControllerEvent(
-              ->setController($controllerName)
-              ->setParam('exception', $exception);
+        $event->setName(MvcEvent::EVENT_DISPATCH_ERROR);
+        $event->setError($application::ERROR_EXCEPTION);
```
Why a variable here btw?

weierophinney replied:
Can't remember why, TBH; that's old code.
On Oct 1, 2015 6:34 AM, "Marco Pivetta" notifications@github.com wrote:

> In src/DispatchListener.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40903533:
> 
> > @@ -211,12 +215,13 @@ protected function marshalBadControllerEvent(
> >          Application $application,
> >          \Exception $exception
> >      ) {
> > -        $event->setError($application::ERROR_EXCEPTION)
> > -              ->setController($controllerName)
> > -              ->setParam('exception', $exception);
> > -        $event->setName(MvcEvent::EVENT_DISPATCH_ERROR);
> > -        $event->setError($application::ERROR_EXCEPTION);
> 
> Why a variable here btw?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40903533.

Ocramius replied:
Alright, will flag up

Ocramius commented on src/DispatchListener.php:
```diff
@@ -211,12 +215,13 @@ protected function marshalBadControllerEvent(
+        $event->setName(MvcEvent::EVENT_DISPATCH_ERROR);
+        $event->setError($application::ERROR_EXCEPTION);
+        $event->setController($controllerName);
+        $event->setParam('exception', $exception);
```
We should probably make this a constant somewhere, so that we don't have users rely on a magic string (and we know when we're breaking BC)

Ocramius commented on src/DispatchListener.php:
```diff
@@ -153,15 +156,16 @@ protected function marshalControllerNotFoundEvent(
+        $event->setName(MvcEvent::EVENT_DISPATCH_ERROR);
+        $event->setError($type);
+        $event->setController($controllerName);
+        $event->setControllerClass('invalid controller class or alias: ' . $controllerName);
```
Eh, weird as heck. Strange that I never noticed this before :|

Ocramius commented on src/Controller/ControllerManager.php:
```diff
@@ -73,10 +74,10 @@ public function injectControllerDependencies($controller, ServiceLocatorInterfac
             // is why the shared EM injection needs to happen; the conditional
             // will always pass.
             $events = $controller->getEventManager();
```
Will need to check this against an interface (as per comment below), as the _aware_ interface defines no getter :-(

weierophinney replied:
Please see my previous comment...
On Oct 1, 2015 6:38 AM, "Marco Pivetta" notifications@github.com wrote:

> In src/Controller/ControllerManager.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40903766:
> 
> > @@ -73,10 +74,10 @@ public function injectControllerDependencies($controller, ServiceLocatorInterfac
> >              // is why the shared EM injection needs to happen; the conditional
> >              // will always pass.
> >              $events = $controller->getEventManager();
> 
> Will need to check this against an interface (as per comment below), as
> the _aware_ interface defines no getter :-(
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40903766.

Ocramius replied:
:+1:

Ocramius commented on src/Service/ApplicationFactory.php:
```diff
@@ -26,6 +26,12 @@ class ApplicationFactory implements FactoryInterface
+        return new Application(
+            $serviceLocator->get('Config'),
+            $serviceLocator,
+            $serviceLocator->get('EventManager'),
```
As mentioned above, it may be a better idea to keep this as it was before

weierophinney replied:
Add noted in my issue summary, I needed to make this change due to a race
condition with EM injection.

Since most people never directly instantiate, this will affect very few
people. I'll be sure to call it out in the migration guide.
On Oct 1, 2015 6:42 AM, "Marco Pivetta" notifications@github.com wrote:

> In src/Service/ApplicationFactory.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40903811:
> 
> > @@ -26,6 +26,12 @@ class ApplicationFactory implements FactoryInterface
> >       */
> >      public function createService(ServiceLocatorInterface $serviceLocator)
> >      {
> > -        return new Application($serviceLocator->get('Config'), $serviceLocator);
> > -        return new Application(
> > -            $serviceLocator->get('Config'),
> > -            $serviceLocator,
> > -            $serviceLocator->get('EventManager'),
> 
> As mentioned above, it may be a better idea to keep this as it was before
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40903811.

Ocramius replied:
Works for me.

Ocramius commented on src/Service/ServiceManagerConfig.php:
```diff
@@ -94,15 +95,20 @@ public function __construct(array $configuration = [])
+                    return;
+                }
+
+                $eventManager = $instance->getEventManager();
```
Same as above. The `aware` interface defines no getter.

weierophinney replied:
For those not following the whole thread, I answered this: `EventManagerAwareInterface` extends `EventManagerCapableInterface`, which defines the getter.

Ocramius commented on src/View/Http/DefaultRenderingStrategy.php:
```diff
@@ -108,9 +108,11 @@ public function render(MvcEvent $e)
-            $events->trigger(MvcEvent::EVENT_RENDER_ERROR, $e);
+
+            $e->setError(Application::ERROR_EXCEPTION);
+            $e->setParam('exception', $ex);
```
Same as above: rely on a constant instead?

weierophinney replied:
Out of scope for this refactor. I am literally trying to do the bare
minimum to get the component working with the new EM version. Open a
separate PR for such changes, if you feel they're important.
On Oct 1, 2015 6:40 AM, "Marco Pivetta" notifications@github.com wrote:

> In src/View/Http/DefaultRenderingStrategy.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40903929:
> 
> > @@ -108,9 +108,11 @@ public function render(MvcEvent $e)
> > 
> > ```
> >          $application = $e->getApplication();
> >          $events      = $application->getEventManager();
> > ```
> > -            $e->setError(Application::ERROR_EXCEPTION)
> > -              ->setParam('exception', $ex);
> > -            $events->trigger(MvcEvent::EVENT_RENDER_ERROR, $e);
> >   +
> > -            $e->setError(Application::ERROR_EXCEPTION);
> > -            $e->setParam('exception', $ex);
> 
> Same as above: rely on a constant instead?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40903929.

Ocramius replied:
Will spawn an issue

Ocramius commented on test/ApplicationTest.php:
```diff
@@ -69,15 +74,13 @@ public function setUp()
-            $callback = $listener->getCallback();
-            if (!is_array($callback)) {
+        $manager = $this->serviceManager->get('ModuleManager');
+        foreach ($this->getListenersForEvent('loadModule', $manager->getEventManager()) as $listener) {
```
Do we have a constant for this event name?

weierophinney replied:
Yep: `ModuleEvent::EVENT_LOAD_MODULE`. I'll update the code.

Ocramius commented on test/EventManagerIntrospectionTrait.php:
```diff
@@ -0,0 +1,56 @@
+ */
+trait EventManagerIntrospectionTrait
+{
+    public function getEventsFromEventManager(EventManager $events)
```
Why can't this be a static method somewhere (instead of a trait)? Also: `private`

weierophinney replied:
I know you hate traits, but I found it an expedient way to accomplish the
task, and this is precisely the dirt of thing they were designed for.

I'll update to private, though I don't think the visibility is important
within a unit test.
On Oct 1, 2015 6:43 AM, "Marco Pivetta" notifications@github.com wrote:

> In test/EventManagerIntrospectionTrait.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40904132:
> 
> > - \* @link      http://github.com/zendframework/zf2 for the canonical source repository
> > - \* @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
> > - \* @license   http://framework.zend.com/license/new-bsd New BSD License
> > - _/
> >   +
> >   +namespace ZendTest\Mvc;
> >   +
> >   +use ReflectionProperty;
> >   +use Zend\EventManager\EventManager;
> >   +
> >   +/_*
> > - \* Offer methods for introspecting event manager events and listeners.
> > - */
> >   +trait EventManagerIntrospectionTrait
> >   +{
> > -    public function getEventsFromEventManager(EventManager $events)
> 
> Why can't this be a static method somewhere (instead of a trait)? Also:
> private
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40904132.

Ocramius replied:
> I know you hate traits

I don't hate traits, I just see a static method as a better solution here. After all, these are helper functions

Ocramius commented on test/EventManagerIntrospectionTrait.php:
```diff
@@ -0,0 +1,56 @@
+        return array_keys($listeners);
+    }
+
+    public function getListenersForEvent($event, EventManager $events, $withPriority = false)
```
Same as above

Ocramius commented on test/EventManagerIntrospectionTrait.php:
```diff
@@ -0,0 +1,56 @@
+        return $this->traverseListeners($listeners[$event], $withPriority);
+    }
+
+    public function traverseListeners(array $queue, $withPriority = false)
```
Same as above

Ocramius replied:
Method purpose is unclear

weierophinney replied:
It's a generator. Suggest a better name?
On Oct 1, 2015 6:44 AM, "Marco Pivetta" notifications@github.com wrote:

> In test/EventManagerIntrospectionTrait.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40904163:
> 
> > -    }
> >   +
> > -    public function getListenersForEvent($event, EventManager $events, $withPriority = false)
> > -    {
> > -        $r = new ReflectionProperty($events, 'events');
> > -        $r->setAccessible(true);
> > -        $listeners = $r->getValue($events);
> >   +
> > -        if (! isset($listeners[$event])) {
> > -            return $this->traverseListeners([]);
> > -        }
> >   +
> > -        return $this->traverseListeners($listeners[$event], $withPriority);
> > -    }
> >   +
> > -    public function traverseListeners(array $queue, $withPriority = false)
> 
> Method purpose is unclear
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40904163.

Ocramius replied:
What does the traversal do? Does it simply allow traversing the listeners by their priority, emulating what the event manager does?

Also, can the argument simply be the event manager (so that fetching the listeners becomes a detail)?

weierophinney replied:
It's the listeners for a specific event. In terms of the priority flag:
- false should be used if you want to potentially cast to an array. An
  interesting thing with generators and iterators is that you can have keys
  that are identical - but iterator_to_array will collapse those. As such,
  this option does not provide keys.
- true should be used if you'll actually iterate, allowing you to get keys
  and listeners.

I could collapse this method into the getListenersForEvent() method,
though; will do so shortly.
On Oct 1, 2015 7:35 AM, "Marco Pivetta" notifications@github.com wrote:

> In test/EventManagerIntrospectionTrait.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40907593:
> 
> > -    }
> >   +
> > -    public function getListenersForEvent($event, EventManager $events, $withPriority = false)
> > -    {
> > -        $r = new ReflectionProperty($events, 'events');
> > -        $r->setAccessible(true);
> > -        $listeners = $r->getValue($events);
> >   +
> > -        if (! isset($listeners[$event])) {
> > -            return $this->traverseListeners([]);
> > -        }
> >   +
> > -        return $this->traverseListeners($listeners[$event], $withPriority);
> > -    }
> >   +
> > -    public function traverseListeners(array $queue, $withPriority = false)
> 
> What does the traversal do? Does it simply allow traversing the listeners
> by their priority, emulating what the event manager does?
> 
> Also, can the argument simply be the event manager (so that fetching the
> listeners becomes a detail)?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40907593.

Ocramius replied:
I'm fine with a descriptive docblock as well, btw

Ocramius commented on test/View/Console/DefaultRenderingStrategyTest.php:
```diff
@@ -18,10 +18,13 @@
-    /** @var DefaultRenderingStrategy */
+    use EventManagerIntrospectionTrait;
+
+    /* @var DefaultRenderingStrategy */
```
Should use `/**`, not `/*`

weierophinney replied:
I saw somebody content recently that these hints shouldn't be formal
docblocks. Can you link me to something showing otherwise?
On Oct 1, 2015 6:44 AM, "Marco Pivetta" notifications@github.com wrote:

> In test/View/Console/DefaultRenderingStrategyTest.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40904191:
> 
> >  class DefaultRenderingStrategyTest extends TestCase
> >  {
> > -    /*\* @var DefaultRenderingStrategy */
> > -    use EventManagerIntrospectionTrait;
> >   +
> > -    /\* @var DefaultRenderingStrategy */
> 
> Should use /*_, not /_
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40904191.

Ocramius replied:
@weierophinney `/**` => docblock (stored in reflection data as well), while `/*` is only for IDE hints.

Here's the practical example: https://3v4l.org/O3CgU

And copying it for reference:

``` php
<?php

class Foo
{
    /* stuff */
    private $bar;
    /** more stuff */
    private $baz;
}

var_dump((new ReflectionProperty(Foo::class, 'bar'))->getDocComment());
var_dump((new ReflectionProperty(Foo::class, 'baz'))->getDocComment());
```

weierophinney replied:
I missed that it was an **object** property. My bad!

Ocramius commented on test/View/RouteNotFoundStrategyTest.php:
```diff
@@ -288,35 +291,33 @@ public function testNotFoundTemplateIsMutable()
-                        $found = true;
-                        break;
-                    }
+            foreach ($listeners as $priority => $listener) {
```
This piece of logic should probably be part of the static/trait above

weierophinney replied:
Working on that now. It's actually an argument for a trait, as then I can do assertions within a method. I'm creating `assertListenerAtPriority(callable $expectedListener, $expectedPriority, $event, EventManager $events, $message = '')` right now; greatly reducing a ton of this boilerplate, and making the test cases easier to understand.

Ocramius commented on test/View/RouteNotFoundStrategyTest.php:
```diff
@@ -16,9 +16,12 @@
 
 class RouteNotFoundStrategyTest extends TestCase
 {
+    use EventManagerIntrospectionTrait;
```
After reading through the tests, I'm more convinced that the trait should just be a static class

weierophinney replied:
Why? What rationale can you provide?
On Oct 1, 2015 6:46 AM, "Marco Pivetta" notifications@github.com wrote:

> In test/View/RouteNotFoundStrategyTest.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40904278:
> 
> >  class RouteNotFoundStrategyTest extends TestCase
> >  {
> > -    use EventManagerIntrospectionTrait;
> 
> After reading through the tests, I'm more convinced that the trait should
> just be a static class
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40904278.

Ocramius replied:
As mentioned above, you are copying the helper methods into every test case, instead of just having them available in the test suite.

Since they are pure helper methods, you can just have them as static methods in a separate class.

weierophinney replied:
Doing so makes the verbiage far more verbose, though:

``` php
$this->getListenersForEvent(...);

// vs
EventManagerIntrospection::getListenersForEvent(...)
```

Let's leaves as-is, with the improvements and additions you've suggested.
I'll get those pushed shortly.
On Oct 1, 2015 7:36 AM, "Marco Pivetta" notifications@github.com wrote:

> In test/View/RouteNotFoundStrategyTest.php
> https://github.com/zendframework/zend-mvc/pull/31#discussion_r40907654:
> 
> >  class RouteNotFoundStrategyTest extends TestCase
> >  {
> > -    use EventManagerIntrospectionTrait;
> 
> As mentioned above, you are copying the helper methods into every test
> case, instead of just having them available in the test suite.
> 
> Since they are pure helper methods, you can just have them as static
> methods in a separate class.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/zendframework/zend-mvc/pull/31/files#r40907654.

Ocramius replied:
Alright

weierophinney committed:
Incorporate feedback

- Updated EventManagerIntrospectionTrait:
  - Made methods private.
  - Added `getArrayOfListenersForEvent()`; casts to array before returning; used
    for counting listeners and/or traversal where priority isn't interesting.
  - Added `assertListenerAtPriority()` for simplifying assertions looking for
    specific listeners registered at specific priorities.
  - Added docblocks to all methods.
- Updated tests per the above.
- Various other small changes as suggested by @Ocramius.

weierophinney commented:
Feedback incorporated, @Ocramius :smile:

weierophinney committed:
CS fixes flagged by php-cs-fixer

weierophinney commented:
Bah, CS checks! Fixed and pushed.

Ocramius commented:
Thanks @weierophinney!

Ocramius commented on composer.json:
```diff
@@ -14,11 +14,12 @@
     "require": {
         "php": ">=5.5",
-        "zendframework/zend-eventmanager": "~2.5",
+        "zendframework/zend-eventmanager": "dev-develop as 2.7.0",
```
Note: should probably use `zend-eventmanager dev-develop as 3.0.0`

weierophinney replied:
Can't yet; not all components are updated to zend-eventmanager v3 yet, which means that there will be an unresolveable conflict. (I know; I tried. :smile:)

weierophinney committed:
Provided initial migration docs

These cover the BC breaks created with the current patch.

weierophinney committed:
Created CHANGELOG for #31

Also updates dev-develop branch to 3.0-dev.

weierophinney commented:
Added a migration guide covering the BC breaks introduced, and updated the branch-alias for the develop branch to 3.0-dev.

weierophinney commented:
Okay, merging. If anybody finds anything else wrong, please open an issue or PR and ping me. :smile:
